core/src/main/scala/kafka/admin/RecommendedLeaderElectionCommand.scala

```diff
@@ -0,0 +1,333 @@
+
+  var preferredBrokers: Set[Int] = _
+
+
+  def main(args: Array[String]): Unit = {
+    run(args, 30.second)
+  }
+
+  def run(args: Array[String], timeout: Duration): Unit = {
+    val commandOptions = new RecommendedLeaderElectionCommandOptions(args)
+    CommandLineUtils.printHelpAndExitIfNeeded(
+      commandOptions,
+      "This tool attempts to elect a new (recommended) leader for a set of topic partitions. To perform a preferred or unclean leader election, please use LeaderElectionCommand instead."
+    )
+
+    validate(commandOptions)
+
+    // This defines the healthy leaders from which to choose; it must exist. Comma-separated list of broker IDs.
+    preferredBrokers = sys.env.get(PREFERRED_BROKER_ENV)
+      .map(_.split(",").toList.map(_.trim.toInt))
+      .map(_.toSet)
+      .getOrElse(throw new IllegalArgumentException(s"Cannot parse env ${PREFERRED_BROKER_ENV}"))
+
+    val jsonFileTopicPartitions: Option[Map[TopicPartition, Integer]] =
+      Option(commandOptions.options.valueOf(commandOptions.pathToJsonFile)).map { path  =>
+        parseReplicaElectionData(Utils.readFileAsString(path))
+      }
+
+    val singleTopicPartition = (
+      Option(commandOptions.options.valueOf(commandOptions.topic)),
+      Option(commandOptions.options.valueOf(commandOptions.partition)),
+      Option(commandOptions.options.valueOf(commandOptions.newLeader)),
+    ) match {
+      case (Some(topic), Some(partition), Some(newLeader)) => Some(Map(new TopicPartition(topic, partition) -> newLeader))
+      case _ => None
+    }
+
+    val topicPartitionsOpt = jsonFileTopicPartitions.orElse(singleTopicPartition)
+
+    val adminClient = {
+      val props = Option(commandOptions.options.valueOf(commandOptions.adminClientConfig)).map { config =>
+        Utils.loadProps(config)
+      }.getOrElse(new Properties())
+
+      props.setProperty(
+        AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG,
+        commandOptions.options.valueOf(commandOptions.bootstrapServer)
+      )
+      props.setProperty(AdminClientConfig.DEFAULT_API_TIMEOUT_MS_CONFIG, timeout.toMillis.toString)
+      props.setProperty(AdminClientConfig.REQUEST_TIMEOUT_MS_CONFIG, (timeout.toMillis / 2).toString)
+
+      Admin.create(props)
+    }
+
+    try {
+      topicPartitionsOpt match {
+        case Some(topicPartitions) =>
+          val total = topicPartitions.size
+          var processed = 0
+          topicPartitions.grouped(100).foreach(partitionSubset => {
+            print(partitionSubset)
+            electRecommendedLeaders(adminClient, partitionSubset)
+            processed += partitionSubset.size
+            if (processed == total) {
+              info(s"Finished processing $processed partitions.")
+            } else {
+              info(s"Processed $processed/$total partitions. Sleeping ${BACK_OFF_DURATION.toSeconds} secs to avoid flooding...\n")
+              Thread.sleep(BACK_OFF_DURATION.toMillis)
+              println(" done")
```

Comment:
nit: should conslidate logging to `debug/info/warn/error`

Reply:
Actually, that's usually a bad idea for command-line tools since the default logger may send output to a random file somewhere. The existing LeaderElectionCommand uses print/println exclusively, and I think this one should follow the same model for both consistency and the Principle of Least Astonishment.

Reply:
Ah that's a good point --- consistency over everything.